File: typescript/0138-copy-list-with-random-pointer.ts
function copyRandomList(head: Node | null): Node | null {
    let hash = new Map<Node, Node>();

    let newHead = new Node(0);

    let itr = head;

    let itrN = newHead;

    while (itr !== null) {
        let newN = new Node(itr.val);
        itrN.next = newN;
        itrN = newN;
        hash.set(itr, newN);

        itr = itr.next;
    }

    itr = head;
    while (itr !== null) {
        let inNew = hash.get(itr);
        if (itr.random !== null) {
            let inNewRand = hash.get(itr.random);
            inNew.random = inNewRand;
        }
        itr = itr.next;
    }

    return newHead.next;
}

File: go/0138-copy-list-with-random-pointer.go
func copyRandomList(head *Node) *Node {
    oldToCopy := make(map[*Node]*Node)
    
    cur := head
    for cur != nil {
        clone := &Node{Val: cur.Val}
        oldToCopy[cur] = clone
        cur = cur.Next
    }
    cur = head
    for cur != nil {
        clone := oldToCopy[cur]
        clone.Next = oldToCopy[cur.Next]
        clone.Random = oldToCopy[cur.Random]
        cur = cur.Next
    }
    return oldToCopy[head]
}
